File: ios/MyApp/Models/NotificationPreference.swift
import Foundation

struct NotificationPreference: Codable, Sendable {
    var morningBriefingEnabled: Bool
    var morningBriefingTime: String
    var deadlineReminderEnabled: Bool
    var deadlineReminderLeadHours: Int
    var overdueAlertsEnabled: Bool
    var fifteenMinuteReminderEnabled: Bool
    var fifteenMinuteReminderLeadMinutes: Int
    var rescheduleAlertsEnabled: Bool

    enum CodingKeys: String, CodingKey {
        case morningBriefingEnabled = "morning_briefing_enabled"
        case morningBriefingTime = "morning_briefing_time"
        case deadlineReminderEnabled = "deadline_reminder_enabled"
        case deadlineReminderLeadHours = "deadline_reminder_lead_hours"
        case overdueAlertsEnabled = "overdue_alerts_enabled"
        case rescheduleAlertsEnabled = "reschedule_alerts_enabled"
        case fifteenMinuteReminderEnabled = "fifteen_minute_reminder_enabled"
        case fifteenMinuteReminderLeadMinutes = "fifteen_minute_reminder_lead_minutes"
    }
}

struct NotificationPreferenceUpdate: Encodable, Sendable {
    var morningBriefingEnabled: Bool?
    var morningBriefingTime: String?
    var deadlineReminderEnabled: Bool?
    var deadlineReminderLeadHours: Int?
    var overdueAlertsEnabled: Bool?
    var fifteenMinuteReminderEnabled: Bool?
    var fifteenMinuteReminderLeadMinutes: Int?
    var rescheduleAlertsEnabled: Bool?

    enum CodingKeys: String, CodingKey {
        case morningBriefingEnabled = "morning_briefing_enabled"
        case morningBriefingTime = "morning_briefing_time"
        case deadlineReminderEnabled = "deadline_reminder_enabled"
        case deadlineReminderLeadHours = "deadline_reminder_lead_hours"
        case overdueAlertsEnabled = "overdue_alerts_enabled"
        case rescheduleAlertsEnabled = "reschedule_alerts_enabled"
        case fifteenMinuteReminderEnabled = "fifteen_minute_reminder_enabled"
        case fifteenMinuteReminderLeadMinutes = "fifteen_minute_reminder_lead_minutes"
    }
}

struct DeviceRegisterRequest: Encodable, Sendable {
    let deviceId: String
    let token: String
    let platform: String
    let timezone: String

    enum CodingKeys: String, CodingKey {
        case deviceId = "device_id"
        case token
        case platform
        case timezone
    }
}

struct DeviceTokenResponse: Codable, Sendable {
    let deviceId: String
    let token: String
    let platform: String
    let timezone: String
    let isActive: Bool

    enum CodingKeys: String, CodingKey {
        case deviceId = "device_id"
        case token
        case platform
        case timezone
        case isActive = "is_active"
    }
}

File: backend/app/models/notification_preference.py
import uuid
from datetime import datetime
from datetime import time as dt_time

from sqlalchemy import Boolean
from sqlalchemy import DateTime
from sqlalchemy import ForeignKey
from sqlalchemy import Integer
from sqlalchemy import Time
from sqlalchemy import UniqueConstraint
from sqlalchemy import func
from sqlalchemy.dialects.postgresql import UUID
from sqlalchemy.orm import Mapped
from sqlalchemy.orm import mapped_column
from sqlalchemy.orm import relationship

from app.db.base import Base


def _default_briefing_time() -> dt_time:
    return dt_time(hour=7, minute=30)


class NotificationPreference(Base):
    __tablename__ = "notification_preferences"
    __table_args__ = (
        UniqueConstraint(
            "user_id",
            name="uq_notification_preferences_user_id",
        ),
    )

    id: Mapped[uuid.UUID] = mapped_column(
        UUID(as_uuid=True),
        primary_key=True,
        default=uuid.uuid4,
    )

    user_id: Mapped[uuid.UUID] = mapped_column(
        UUID(as_uuid=True),
        ForeignKey("users.id", ondelete="CASCADE"),
        nullable=False,
        index=True,
    )

    morning_briefing_enabled: Mapped[bool] = mapped_column(
        Boolean,
        nullable=False,
        default=True,
    )

    morning_briefing_time: Mapped[dt_time] = mapped_column(
        Time,
        nullable=False,
        default=_default_briefing_time,
    )

    deadline_reminder_enabled: Mapped[bool] = mapped_column(
        Boolean,
        nullable=False,
        default=True,
    )

    deadline_reminder_lead_hours: Mapped[int] = mapped_column(
        Integer,
        nullable=False,
        default=24,
    )

    overdue_alerts_enabled: Mapped[bool] = mapped_column(
        Boolean,
        nullable=False,
        default=True,
    )

    fifteen_minute_reminder_enabled: Mapped[bool] = mapped_column(
        Boolean,
        nullable=False,
        default=True,
    )

    fifteen_minute_reminder_lead_minutes: Mapped[int] = mapped_column(
        Integer,
        nullable=False,
        default=15,
    )

    reschedule_alerts_enabled: Mapped[bool] = mapped_column(
        Boolean,
        nullable=False,
        default=True,
    )

    created_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True),
        server_default=func.now(),
        nullable=False,
    )

    updated_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True),
        server_default=func.now(),
        onupdate=func.now(),
        nullable=False,
    )

    user: Mapped["User"] = relationship(back_populates="notification_preferences")
